Bad suggestion when calling `iter` instead of `iter_mut`

Given:

```rust
use std::collections::HashMap;

struct Test {
    v: u32,
}

fn main() {
    let mut map = HashMap::new();
    map.insert("a", Test { v: 0 });

    for (k, mut v) in map.iter() {
        v.v += 1;
    }
}
```

we emit:

```
error[E0594]: cannot assign to `v.v` which is behind a `&` reference
  --> src/main.rs:12:9
   |
11 |     for (k, mut v) in map.iter() {
   |             ----- help: consider changing this to be a mutable reference: `&mut Test`
12 |         v.v += 1;
   |         ^^^^^^^^ `v` is a `&` reference, so the data it refers to cannot be written
```

but we should be instead suggesting

```
error[E0594]: cannot assign to `v.v` which is behind a `&` reference
  --> src/main.rs:12:9
   |
11 |     for (k, mut v) in map.iter() {
   |             -----         ---- help: consider calling `iter_mut` instead
   |             |
   |             this binding is of type `&Test`
12 |         v.v += 1;
   |         ^^^^^^^^ `v` is a `&` reference, so the data it refers to cannot be written
```

CC #49839

--- a/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
@@ -787,7 +787,12 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                                                         _,
                                                         [
                                                             Expr {
-                                                                kind: MethodCall(path_segment, ..),
+                                                                kind:
+                                                                    MethodCall(
+                                                                        path_segment,
+                                                                        _args,
+                                                                        span,
+                                                                    ),
                                                                 hir_id,
                                                                 ..
                                                             },
@@ -831,7 +836,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                 if let Some(mut suggestions) = opt_suggestions {
                     if suggestions.peek().is_some() {
                         err.span_suggestions(
-                            path_segment.ident.span,
+                            *span,
                             "use mutable method",
                             suggestions,
                             Applicability::MaybeIncorrect,

--- a/src/test/ui/suggestions/suggest-mut-method-for-loop-hashmap.fixed
+++ b/src/test/ui/suggestions/suggest-mut-method-for-loop-hashmap.fixed
@@ -1,0 +1,21 @@
+// run-rustfix
+// https://github.com/rust-lang/rust/issues/82081
+
+use std::collections::HashMap;
+
+struct Test {
+    v: u32,
+}
+
+fn main() {
+    let mut map = HashMap::new();
+    map.insert("a", Test { v: 0 });
+
+    for (_k, mut v) in map.iter_mut() {
+        //~^ HELP use mutable method
+        //~| NOTE this iterator yields `&` references
+        v.v += 1;
+        //~^ ERROR cannot assign to `v.v`
+        //~| NOTE `v` is a `&` reference
+    }
+}

--- a/src/test/ui/suggestions/suggest-mut-method-for-loop-hashmap.rs
+++ b/src/test/ui/suggestions/suggest-mut-method-for-loop-hashmap.rs
@@ -1,0 +1,21 @@
+// run-rustfix
+// https://github.com/rust-lang/rust/issues/82081
+
+use std::collections::HashMap;
+
+struct Test {
+    v: u32,
+}
+
+fn main() {
+    let mut map = HashMap::new();
+    map.insert("a", Test { v: 0 });
+
+    for (_k, mut v) in map.iter() {
+        //~^ HELP use mutable method
+        //~| NOTE this iterator yields `&` references
+        v.v += 1;
+        //~^ ERROR cannot assign to `v.v`
+        //~| NOTE `v` is a `&` reference
+    }
+}

--- a/src/test/ui/suggestions/suggest-mut-method-for-loop-hashmap.stderr
+++ b/src/test/ui/suggestions/suggest-mut-method-for-loop-hashmap.stderr
@@ -1,0 +1,15 @@
+error[E0594]: cannot assign to `v.v`, which is behind a `&` reference
+  --> $DIR/suggest-mut-method-for-loop-hashmap.rs:17:9
+   |
+LL |     for (_k, mut v) in map.iter() {
+   |                        ----------
+   |                        |   |
+   |                        |   help: use mutable method: `iter_mut()`
+   |                        this iterator yields `&` references
+...
+LL |         v.v += 1;
+   |         ^^^^^^^^ `v` is a `&` reference, so the data it refers to cannot be written
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0594`.

